feat: yaml patch override

dannyrandall opened the pull request:
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

dannyrandall committed:
parse patches

dannyrandall committed:
fix: override rdws templates even if alias is set

dannyrandall committed:
added testing code

dannyrandall committed:
add to map appears to work....so...

dannyrandall committed:
saving some stuff

dannyrandall committed:
new algo for add

dannyrandall committed:
remove works

dannyrandall committed:
replace works

dannyrandall committed:
wrote v1 of helper func

dannyrandall committed:
reusable pointer traversing func

dannyrandall committed:
change function names

dannyrandall committed:
license

dannyrandall committed:
Merge branch 'mainline' into feat/yaml-patch

efekarakus commented on internal/pkg/cli/deploy/rdws.go:
```diff
@@ -176,7 +176,7 @@ func (d *rdwsDeployer) stackConfiguration(in *StackRuntimeConfiguration) (*rdwsS
 	return &rdwsStackConfigurationOutput{
 		svcStackConfigurationOutput: svcStackConfigurationOutput{
-			conf: conf,
+			conf: cloudformation.WrapWithTemplateOverrider(conf, d.overrider),
```
:doge: was this a miss from my end?

dannyrandall replied:
maybe 🙄

efekarakus commented on internal/pkg/override/patch.go:
```diff
@@ -0,0 +1,253 @@
+		case "replace":
+			err = patch.applyReplace(&root)
+		default:
+			return nil, fmt.Errorf("unsupported operation %q", patch.Operation)
```
nit: should we including the valid set of operations as hint here? perhaps with `RecommendActions` 💭

efekarakus commented on internal/pkg/override/patch.go:
```diff
@@ -0,0 +1,253 @@
+	fs          afero.Fs // OS file system.
+}
+
+type PatchOpts struct {
```
Doc comments missing for `PatchOpts`, `WithPatch` and `Override` :)

efekarakus commented on internal/pkg/override/patch.go:
```diff
@@ -0,0 +1,253 @@
+		case "add":
+			err = patch.doAdd(&root)
+		case "remove":
+			err = patch.doRemove(&root)
```
nit: what do you think of renaming this like 👇  to be consistent w 
```go
	err = patch.applyAdd(&root)
case "remove":
	err = patch.applyRemove(&root)
```
Alternatively, we can pull out the `visit` functions and do something like:
```go
	err = patch.apply(&root, addLeafFn)
case "remove":
	err = patch.apply(&root, removeLeafFn)
```
where 
```go
func addLeadFn(node *yaml.Node, pointer []string) error { ... }
```

efekarakus commented on internal/pkg/override/patch.go:
```diff
@@ -0,0 +1,253 @@
+	switch node.Kind {
+	case yaml.DocumentNode:
+		if len(node.Content) != 1 {
+			return fmt.Errorf("don't support multi-doc yaml")
```
What does this mean 😅

dannyrandall replied:
you can [embed two yaml documents in a single file](https://yaml.org/spec/1.2.2/#22-structures). since multi document files don't exist in JSON (and therefore JSON Pointer syntax), there is no way to target a specific document with a JSON Pointer. So we just assume the target is the first document, and error if there are multiple documents in the file.

dannyrandall replied:
update: https://pkg.go.dev/gopkg.in/yaml.v3#Unmarshal 🤦 
> Unmarshal decodes the first document

Since we only call unmarshal once, I think it's ok if we just ignore this case.

efekarakus commented on internal/pkg/override/patch.go:
```diff
@@ -0,0 +1,253 @@
+	split := strings.Split(pointer, "/")
+	for i := range split {
+		split[i] = strings.ReplaceAll(split[i], "~0", "~")
+		split[i] = strings.ReplaceAll(split[i], "~1", "/")
```
nit: Would you mind adding a comment to section 3. of https://www.rfc-editor.org/rfc/rfc6901 to explain these characters :)

They also recommend doing the opposite sequence 😂 
> This is performed by first transforming any occurrence of the sequence '\~1' to '/', and then transforming any occurrence of the sequence '\~0' to '\~'.

efekarakus commented on internal/pkg/override/patch.go:
```diff
@@ -0,0 +1,253 @@
+		case yaml.MappingNode:
+			for i := 0; i < len(node.Content); i += 2 {
+				if node.Content[i].Value == pointer[0] {
+					node.Content = append(node.Content[:i], node.Content[i+2:]...)
```
will this work if we try to remove the last key in a map? (i.e. can we do `node.Content[i+2:]`?)

dannyrandall replied:
yeah - the last key in the map should be at index `len(node.Content)-2`, so the index `i+2` would be `len(node.Content)` - which, when slicing a slice, produces a slice of length 0. (also see [this test](https://github.com/aws/copilot-cli/pull/4641/files#diff-db45db57a8c5fa7983bfd1bd51464cb707a656831be2aa65c72bf3f758cf296cR172-R187))

```
func main() {
  list := []string{"a", "b", "c"}
  fmt.Printf("list: %+v, len(): %d\n", list, len(list))
  fmt.Printf("list[3:] %+v\n", list[3:])
}

// Output:
// list: [a b c], len(): 3
// list[3:] []
```

efekarakus commented on internal/pkg/override/patch.go:
```diff
@@ -0,0 +1,253 @@
+func (y *yamlPatch) doRemove(root *yaml.Node) error {
+	return followJSONPointer(root, y.Path, func(node *yaml.Node, pointer []string) error {
+		if len(pointer) != 1 {
+			return nil
```
I wonder if the `nil` and `errStopFollowingPointer` pointer relationship should be reversed. For example,

```go
if len(pointer) > 1 {
   return errSkip
}
// ...
return nil  // if successfully added, removed, or replaced an element.
```
(inspired by https://cs.opensource.google/go/go/+/refs/tags/go1.20.2:src/io/fs/walk.go;l=15)

dannyrandall replied:
i think the relationship is the same here as `io/fs` - `Walk` continues calling the `WalkDirFunction` on every file until `SkipDir`, `SkipAll`, or a non-nil error is returned... We keep calling `visit()` until `errStopFollowingPointer` or a non-nil error is returned. do you think the naming is potentially confusing...?

efekarakus commented on internal/pkg/override/patch.go:
```diff
@@ -0,0 +1,253 @@
+	return followJSONPointerHelper(root, split, visit)
+}
+
+func followJSONPointerHelper(node *yaml.Node, pointer []string, visit func(node *yaml.Node, pointer []string) error) error {
```
I wonder if the `visit` function should take as input the subpath traveled so far that in our errors we can provide exactly where the issue occurs?
```suggestion
func followJSONPointerHelper(node *yaml.Node, pointer []string, visit func(node *yaml.Node, traversed, remaining []string) error) error {
```
so that we can do something like:
```go
if err := visit(node, pointer); err != nil && !errors.Is(err, errSkip) {
   return fmt.Errorf("visit %q: %v", strings.Join(traversed, jsonPathSeparator), err)
}

// and further down
return fmt.Errorf("visit %q: key %q not found in map", strings.Join(traversed, jsonPathSeparator),pointer[0])
```

efekarakus commented on internal/pkg/override/patch.go:
```diff
@@ -0,0 +1,253 @@
+		return nil, fmt.Errorf("read directory %q: %w", p.rootAbsPath, err)
+	}
+
+	for _, file := range files {
```
What do you think of just expecting a _single file_ for now under the directory that conforms to the YAML patch format (otherwise returning an error)?
I think in the ideal world, we will have `copilot [noun] override --tool yamlpatch` generate two files under the `overrides` directory: `README.md` and `cfn.patches.yaml`.

We might need to also update to no longer expect a single file but if there is no `cdk.json` then assume it's YAML patch.
https://github.com/aws/copilot-cli/blob/caa6aca43b9ab4f0df4b80430fd36eebc532d78a/internal/pkg/override/override.go#L87-L91

dannyrandall replied:
sure, a single file sounds good to me - people can use comments in that file to help organize their overrides. i'll make that change!

efekarakus reviewed (commented):
Looks really good, it was really easy to follow 🤩 

Mostly minor recommendations 🚀

dannyrandall committed:
address feedback from @efekarakus:
- comments on exported values
- include traversed path in error messages
- swap order of text replacements

dannyrandall committed:
more feedback from @efekarakus:
- consistent `apply` naming

dannyrandall committed:
more feedback (@efekarakus): only support a single file

dannyrandall committed:
support `-` in middle of pointer

dannyrandall commented:
I've been looking into reusing/merging some code from `taskdef_overrides` - there is one major difference in how patches work vs. taskdef overrides that complicates that: taskdef overrides recursively creates nodes from the input _path_, whereas yaml patches require a pointer that points to an existing (or new leaf, in the case of "add") node in the tree.

For example, given the input:
```
map:
  key: value
```
This works in taskdef overrides:
```
path: map.newKey.subKey
value: subKey_value
```
This not supported in a YAML Patch per [RFC6902](https://www.rfc-editor.org/rfc/rfc6902). Specifically:
> However, the object itself or an array containing it does need to exist, and it remains an error for that not to be the case.  For example, an "add" with a target location of "/a/b" starting with this document:
>
>   { "a": { "foo": 1 } }
>
>  is not an error, because "a" exists, and "b" will be added to its value.  It is an error in this document:
>
>   { "q": { "bar": 2 } }
>
>   because "a" does not exist.

To accomplish the same thing from my example in yaml patches, the operation would look like:
```
op: add
path: map/newKey
value:
  subKey: subKey_value
```
---
At this point, I'm hesitant to merge the two logical paths (one creating nodes as it follows a path, and one that fails) in favor of ensuring compatibility for existing taskdef override users. I think in a future release we can instead help customers migrate from taskdef overrides to yaml patches with some logic to convert an existing override to a patch - and as part of that logic, merge the two patterns by applying the conversion before applying the patches.

dannyrandall committed:
add operation should only append when using `-`

dannyrandall committed:
fix document node handling, add couple of error test cases

dannyrandall committed:
test coverage

dannyrandall committed:
Merge branch 'mainline' into feat/yaml-patch

efekarakus commented on internal/pkg/override/patch.go:
```diff
@@ -0,0 +1,268 @@
+		case "replace":
+			err = patch.applyReplace(&root)
+		default:
+			return nil, fmt.Errorf("unsupported operation %q. supported operations are %q, %q, and %q.", patch.Operation, "add", "remove", "replace")
```
nit:
```suggestion
			return nil, fmt.Errorf("unsupported operation %q: supported operations are %q, %q, and %q.", patch.Operation, "add", "remove", "replace")
```

efekarakus commented on internal/pkg/override/patch.go:
```diff
@@ -0,0 +1,268 @@
+
+		return fmt.Errorf("key %q: %q not found in map", strings.Join(traversed, jsonPointerSeparator), remaining[0])
+	case yaml.SequenceNode:
+		if remaining[0] == "-" {
+			return followJSONPointerHelper(node.Content[len(node.Content)-1], append(traversed, remaining[0]), remaining[1:], visit)
+		}
```
ooh would you mind adding a comment here saying that removing elements using "-" is not allowed according to the spec:
> Note that the use of the "-" character to index an array will always
   result in such an error condition because by definition it refers to
   a nonexistent array element.  

I initially thought that doing the following was a valid option:
```go
- op: remove
  path: /Resources/TaskDef/List/-
```

dannyrandall replied:
hopefully the updated logic is sufficient; `-` logic only exists in `add` now

efekarakus commented on internal/pkg/override/patch_test.go:
```diff
@@ -0,0 +1,452 @@
+- op: add
+  path: /a/b/1/e
+  value: val`,
+			expectedErr: `unable to apply "add" patch: key "/a/b/1": invalid node type 0x8`,
```
😅  the node type `0x8` won't make sense to users, can we add a `nodeKindStringer` type to make the error more friendly?
```go
type nodeKindStringer yaml.Kind

// String implements fmt.Stringer.
func (k nodeKindStringer) String() string {
   switch k {
   case yaml.DocumentNode:
      return "document"
   case yaml.MappingNode:
      return "mapping"
   // ...
   }
}
```

and in `followJSONPointerHelper`:
```go
return fmt.Errorf("key %q: invalid node type %q", strings.Join(traversed, jsonPointerSeparator), nodeKindStringer(node.Kind))
```

efekarakus reviewed (approved):
Some error message suggestions, feel free to remove the label afterwards

dannyrandall committed:
address nit fb from @efekarakus:
- consistent error format
- node kind string in error messages

dannyrandall committed:
allow adding to end of list by index

dannyrandall committed:
remove local test

dannyrandall committed:
only support `-` at end of `add` op path

dannyrandall committed:
error test case with `-` in middle of seq

dannyrandall committed:
Merge branch 'mainline' into feat/yaml-patch

Lou1415926 commented on internal/pkg/override/patch.go:
```diff
@@ -0,0 +1,283 @@
+	}
+}
+
+type yamlPatch struct {
```
nit: can we move this type declaration to just before `func (p *yamlPatch) applyRemove(root *yaml.Node) error` ?

Lou1415926 commented on internal/pkg/override/patch.go:
```diff
@@ -0,0 +1,283 @@
+			return nil, fmt.Errorf("unsupported operation %q: supported operations are %q, %q, and %q.", patch.Operation, "add", "remove", "replace")
+		}
+		if err != nil {
+			return nil, fmt.Errorf("unable to apply %q patch: %w", patch.Operation, err)
```
what do you think of including the index information? So that people see for example `unable to apply the patch at index 0 with "add": value required`

dannyrandall replied:
good idea!!

Lou1415926 commented on internal/pkg/override/patch.go:
```diff
@@ -0,0 +1,283 @@
+		return fmt.Errorf("key %q: %w", strings.Join(traversed, jsonPointerSeparator), err)
+	} else if len(remaining) == 0 {
+		return nil
+	}
```
nit: this seems to be the same?
```suggestion
	}
	if len(remaining) == 0 {
		return nil
	}
```

dannyrandall replied:
code is gone in refactor!

Lou1415926 commented on internal/pkg/override/patch.go:
```diff
@@ -0,0 +1,283 @@
+	switch node.Kind {
+	case yaml.DocumentNode:
+		if len(node.Content) == 0 {
+			return nil // weird, but ok ¯\_(ツ)_/¯
```
When would this happen? Maybe only when the first document is empty? What about something like ⬇️  😂 

```
// NOTE: happens only when xxx; realistically wouldn't happen (weird, but ok ¯\_(ツ)_/¯)
```

dannyrandall replied:
i've played with it for a while and can't actually find a yaml document that would make this occur - i'll add a note that it's just there to avoid panics. In theory it would be something like 
```
---
key: value
```
but the parser still creates a document node in this case, with a `!!null` node inside.

dannyrandall replied:
i'll switch it to spit out an error as well

Lou1415926 commented on internal/pkg/override/patch.go:
```diff
@@ -0,0 +1,283 @@
+			return nil
+		}
+
+		switch node.Kind {
```
How does the `ScalarNode` case play with the this? I guess it'd error out in `followHelper` because if the JSON pointer goes to a scalar node, it doesn't make sense for "add"?

dannyrandall replied:
yeah previously this was caught by the followHelper, but in the refactor i now have a case for scalar nodes here!

Lou1415926 commented on internal/pkg/override/patch_test.go:
```diff
@@ -0,0 +1,483 @@
+	"gopkg.in/yaml.v3"
+)
+
+func TestPatch(t *testing.T) {
```
noice  test cases👍🏼

Lou1415926 commented on internal/pkg/override/patch.go:
```diff
@@ -0,0 +1,283 @@
+		}
+
+		switch node.Kind {
+		case yaml.DocumentNode:
```
How would we hit this case?

dannyrandall replied:
it's a fairly...useless scenario for our use case, but with the following patch:
```
op: remove
path: ""
```
(see this test https://github.com/aws/copilot-cli/pull/4641/files#diff-db45db57a8c5fa7983bfd1bd51464cb707a656831be2aa65c72bf3f758cf296cR344-R350)

dannyrandall committed:
refactor to remove `visit()` pattern, add a couple of tests for errors

dannyrandall committed:
fb from @Lou1415926: include patch idx in error

dannyrandall committed:
addr nits from @Lou1415926:
- move `yamlPatch` struct
- error in 0-content doc case

iamhopaul123 commented on internal/pkg/override/patch.go:
```diff
@@ -0,0 +1,318 @@
+		return err
+	}
+
+	switch parent.Kind {
+	case yaml.DocumentNode:
+		return parent.Encode(p.Value)
+	case yaml.MappingNode:
+		i, err := findInMap(parent, pointer.FinalKey(), pointer.Parent())
+		if err == nil {
+			// if the key is in this map, they are trying to replace it
+			return parent.Content[i+1].Encode(p.Value)
+		}
+
+		// if the key isn't in this map, then we need to create it for them
+		parent.Content = append(parent.Content, &yaml.Node{
+			Kind:  yaml.ScalarNode,
+			Tag:   "!!str",
+			Value: pointer.FinalKey(),
+		})
+		parent.Content = append(parent.Content, &p.Value)
+	case yaml.SequenceNode:
+		if pointer.FinalKey() == "-" {
+			// add to end of sequence
+			parent.Content = append(parent.Content, &p.Value)
+			return nil
```
Is this possible to make this whole thing `parent.add(p.Value)`? sorry i can't include more lines but it should reach to the end.

and `applyAdd(root *yaml.Node)` will be clearer: we find the target node to add/replace/delete, and then apply the action.

dannyrandall replied:
problem is that replace doesn't use the parent node, it uses the "node" node - so depending on the action, we need a different node

dannyrandall replied:
though i guess the action for replace could be - take the parent, get the correct child node using `getNode()` and replace that?

iamhopaul123 replied:
no i just meant we can wrap this up to make the logic clearer. It can be different for replace doesn't necessarily need to be `parent.`

dannyrandall replied:
i think all of the logic for each operation (add/remove/replace) is totally separate, so i'm not sure there's much benefit in breaking out the find parent and "do operation" logic, since it can't get reused in any way...

iamhopaul123 commented on internal/pkg/override/patch.go:
```diff
@@ -0,0 +1,318 @@
+	return p[len(p)-1]
+}
+
+func (y yamlPatch) Pointer() pointer {
```
Why do we make them public?

iamhopaul123 commented on internal/pkg/override/patch.go:
```diff
@@ -0,0 +1,318 @@
+	return 0, fmt.Errorf("key %q: %q not found in map", strings.Join(traversed, jsonPointerSeparator), key)
+}
+
+func getNode(node *yaml.Node, remaining, traversed pointer) (*yaml.Node, error) {
```
maybe consider making the name more meaningful since it is a crucial one. Like `findTargetNode` or sth similar.

iamhopaul123 reviewed (commented):
LGTM overall! just few nits

dannyrandall committed:
fb from @iamhopaul123: make functions private

dannyrandall committed:
Merge branch 'mainline' into feat/yaml-patch

dannyrandall committed:
Merge branch 'mainline' into feat/yaml-patch
